### ISSUE TITLE
Getting a high resoloution image of a large PDF causes a segmentation fault

### ISSUE THREAD

**Issue by dbrnz** [pymupdf/PyMuPDF#675]
Attempting to get a high-resolution image of a large PDF leads to malloc failures and seg faults.

To reproduce:
1) Create a PDF consisting of an entirely blank page, 120cm x 90cm, as attached ([blank.pdf](https://github.com/pymupdf/PyMuPDF/files/5314671/blank.pdf)). This was done by exporting a blank slide from Powerpoint. 
2) Run the following:
```
import sys
import fitz

def get_image(scaling):
    print('Get image with scaling', scaling)
    try:
        pixmap = page.getPixmap(matrix=fitz.Matrix(scaling, scaling), alpha=True, clip=None)
        print('  Got pixmap: size: {}, stride: {}, (width, height): {}'
              .format(pixmap.size, pixmap.stride, (pixmap.width, pixmap.height)))
        data = pixmap.getImageData('png')
        print('  Got image data...')
    except RuntimeError as error:
        print(error)
    print('')

print(sys.version, '\n', sys.platform, '\n', fitz.__doc__)

pdf = fitz.Document('blank.pdf', filetype='application/pdf')

page = pdf[0]
print('PDF size:', page.rect, '\n')

get_image(7)
get_image(8)
get_image(12)
```

This results in:
```
3.7.5 (default, Nov 29 2019, 12:44:14)
[Clang 11.0.0 (clang-1100.0.33.12)]
 darwin

PyMuPDF 1.17.7: Python bindings for the MuPDF 1.17.0 library.
Version date: 2020-09-14 06:33:06.
Built for Python 3.7 on darwin (64-bit).

PDF size: Rect(0.0, 0.0, 3402.0, 2551.0)

Get image with scaling 7
  Got pixmap: size: 1700986480, stride: 95256, (width, height): (23814, 17857)
  Got image data...

Get image with scaling 8
  Got pixmap: size: -2073270696, stride: 108864, (width, height): (27216, 20408)
mupdf: malloc of 18446744071636280832 bytes failed
malloc of 18446744071636280832 bytes failed

Get image with scaling 12
  Got pixmap: size: 703849944, stride: 163296, (width, height): (40824, 30612)
[1]    6804 segmentation fault  python test.py
```

`pyMuPDF` was installed from PyPI.

**Comment by JorjMcKie** [comment]
The malloc failures are okay: they *must* happen if out of  memory.

Bringing down the interpreter (segmentation faults or whatever) is however never okay.

I was able to reproduce your error on Linux (Windows WSL), but not on Windows …
This is an upstream error (MuPDF). If an alloc fails, MuPDF tries to free memory from its  cache and recycles (process is called “scavenging”). Something must be going wrong there.
Fortunately I am forcing MuPDF to use Python memory allocation, so I may be able to prevent the scavenging from happening and issue an out of memory exception myself if allocation does not work on first try.

**Comment by JorjMcKie** [comment]
Just implemented a fix as indicated in my previous post: seems to work!
I believe it should become part of the next PyMuPDF version. To recap what it does:

  1.  Provide an interface for MuPDF to acquire memory by using Python functions for this purpose.
  2.  Do not return unsuccessful memory acuisitions to MuPDF, but instead immediately generate an exception. This prevents MuPDF from  scavenging.

**Comment by dbrnz** [comment]
Besides the malloc failures and seg faults in `getImageData()` something strange is going on in `getPixmap()`. Looking at a pixmap's reported size/stride/width/height, and also running for scaling 1 and 4, we get:
```
PDF size: Rect(0.0, 0.0, 3402.0, 2551.0)

Get image with scaling 1
  Got pixmap: size: 34714096, stride: 13608, (width, height): (3402, 2551)
  Got image data...

Get image with scaling 4
  Got pixmap: size: 555424216, stride: 54432, (width, height): (13608, 10204)
  Got image data...

Get image with scaling 7
  Got pixmap: size: 1700986480, stride: 95256, (width, height): (23814, 17857)
  Got image data...

Get image with scaling 8
  Got pixmap: size: -2073270696, stride: 108864, (width, height): (27216, 20408)
mupdf: malloc of 18446744071636280832 bytes failed
malloc of 18446744071636280832 bytes failed
```
For scaling of 1, 4, and 7, `stride` is found to be `4*width` and `size` to be `stride*height + 88`. Size should then be 2,221,696,600 for the scaling 8 case, instead of the reported -2,073,270,696. Adding 2,221,696,600 and 2,073,270,696 gives exactly 2^32 -- is MuPDF using 32-bit `int` instead of 64-bit `long` for its variables??

**Comment by dbrnz** [comment]
> I was able to reproduce your error on Linux (Windows WSL), but not on Windows …

Has MuPDF been built under Windows with a (default?) compiler option that forces `int` to be 64-bit whereas the Linux/macOS builds are defaulting to 32-bit? Or are the relevant variables in fact `long` but the compiler is defaulting to 32-bit longs??

**Comment by JorjMcKie** [comment]
Yes: the pixmap values size, width, height and stride are size_t in MuPDF which I until now clamped to C integers. This shouldn’t be a problem except for crazy pixmap sizes.
Anyway, I already changed it to Py_ssize_t together with the other fix.

**Comment by JorjMcKie** [comment]
There never was a difference between Windows and Linux as far as the integer overflow is concerned: both platforms showed the same wrong pixmap sizes for the very large pixmaps.
``malloc`` messages appear under Windows, too, but I never saw an interpreter crash there. This may simply have been good luck: in your environment there also appeared some malloc exceptions (which are good) before the interpreter crashed because of segfaults.
After replacing ``int`` by ``Py_ssize_t`` the pixmap sizes now are correct on all platforms.

**Comment by JorjMcKie** [comment]
This issue is fixed with v1.18.0.

**Comment by dbrnz** [comment]
Thanks!
